File: 0001-0500/0437. Path Sum III.go
package main
// Medium
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/path-sum-iii/

// We don't need to start from root, could start from any node, so use prefixSum to store

func pathSum(root *TreeNode, targetSum int) int {
    prefixSum := make(map[int]int)
    prefixSum[0] = 1
    return dfs(root, 0, targetSum, prefixSum)
}

func dfs(node *TreeNode, curSum int, targetSum int, prefixSum map[int]int) int {
    if node == nil {
        return 0
    }

    curSum += node.Val
    count := prefixSum[curSum - targetSum]
    prefixSum[curSum]++

    count += dfs(node.Left, curSum, targetSum, prefixSum)
    count += dfs(node.Right, curSum, targetSum, prefixSum)

    prefixSum[curSum]--
    return count
}

File: 0001-0500/0437. Path Sum III.py
# Time:O(n), Space:O(n)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        prefix_sum = {0: 1}

        def dfs(node, cur_sum):
            if not node:
                return 0

            cur_sum += node.val
            count = 0

            if cur_sum - targetSum in prefix_sum:
                count = prefix_sum[cur_sum - targetSum]

            prefix_sum[cur_sum] = prefix_sum.get(cur_sum, 0) + 1

            count += dfs(node.left, cur_sum)
            count += dfs(node.right, cur_sum)

            prefix_sum[cur_sum] -= 1

            return count

        return dfs(root, 0)
